pkg/ds/daemon_set.go

```diff
@@ -0,0 +1,353 @@
+func (ds *daemonSet) addPods() error {
+	currentNodes, err := ds.CurrentNodes()
+	if err != nil {
+		return util.Errorf("Error retrieving current nodes daemon set: %v", err)
+	}
+
+	eligible, err := ds.scheduler.EligibleNodes(ds.Manifest, ds.NodeSelector)
+	if err != nil {
+		return util.Errorf("Error retrieving eligible nodes for daemon set: %v", err)
+	}
+
+	// Get the difference in nodes that we need to schedule on and then sort them
+	// for deterministic ordering
+	toScheduleSorted := types.NewNodeSet(eligible...).Difference(types.NewNodeSet(currentNodes...)).ListNodes()
+	ds.logger.NoFields().Infof("Need to schedule %d nodes", len(toScheduleSorted))
+
+	for _, node := range toScheduleSorted {
+		err := ds.schedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+// removePods unschedules pods for all scheduled nodes not selected
+// by ds.nodeSelector
+func (ds *daemonSet) removePods() error {
+	currentNodes, err := ds.CurrentNodes()
+	if err != nil {
+		return util.Errorf("Error retrieving current nodes daemon set: %v", err)
+	}
+
+	eligible, err := ds.scheduler.EligibleNodes(ds.Manifest, ds.NodeSelector)
+	if err != nil {
+		return util.Errorf("Error retrieving eligible nodes for daemon set: %v", err)
+	}
+
+	// Get the difference in nodes that we need to unschedule on and then sort them
+	// for deterministic ordering
+	toUnscheduleSorted := types.NewNodeSet(currentNodes...).Difference(types.NewNodeSet(eligible...)).ListNodes()
+	ds.logger.NoFields().Infof("Need to unschedule %d nodes", len(toUnscheduleSorted))
+
+	for _, node := range toUnscheduleSorted {
+		err := ds.unschedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+// clearPods unschedules pods for all the nodes that have been scheduled by
+// this daemon set by using CurrentNodes()
+// This should only be used when a daemon set is deleted
+func (ds *daemonSet) clearPods() error {
+	currentNodes, err := ds.CurrentNodes()
+	if err != nil {
+		return util.Errorf("Error retrieving current nodes daemon set: %v", err)
+	}
+
+	// Get the difference in nodes that we need to unschedule on and then sort them
+	// for deterministic ordering
+	toUnscheduleSorted := types.NewNodeSet(currentNodes...).ListNodes()
+	ds.logger.NoFields().Infof("Need to unschedule %d nodes", len(toUnscheduleSorted))
+
+	for _, node := range toUnscheduleSorted {
+		err := ds.unschedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+func (ds *daemonSet) schedule(node types.NodeName) error {
+	ds.logger.NoFields().Infof("Scheduling on %s", node)
+
+	// Will apply the following label on the key <labels.NODE>/<node.String()>:
+	// 	{ DSIDLabel : ds.ID().String() }
+	// eg node/127.0.0.1[daemon_set_id] := test_ds_id
+	err := ds.nodeApplicator.SetLabel(labels.NODE, node.String(), DSIDLabel, ds.ID().String())
```

Comment:
i suspect the latter, because there might be two daemon sets scheduling different pods on the same node, and they would overwrite each other's labels